util/src/MultiSizeBufferPool.cpp

```diff
@@ -0,0 +1,687 @@
+      throwException<std::invalid_argument>(logger_,
+                                            "Invalid pool subpools configuration: bufferSize must be ascending!",
+                                            KVLOG(KVLOG_PREFIX, subpoolConfig.bufferSize, lastBufferSize.value()));
+    }
+    lastBufferSize = subpoolConfig.bufferSize;
+    bufferSizes_.insert(lastBufferSize.value());
+    subPoolName = name_ + "_" + to_string(lastBufferSize.value());
+
+    auto [_, insertFlag] =
+        subpools_.emplace(std::make_pair(lastBufferSize.value(),
+                                         std::make_unique<MultiSizeBufferPool::SubPool>(
+                                             logger_, subpoolConfig, std::move(subPoolName), config_, stats_)));
+    UNUSED(_);
+    ConcordAssert(insertFlag == true);
+  }  // for
+
+  // Set limits on statistics
+  stats_.setLimits(config_.maxAllocatedBytes);
+
+  if ((config_.purgeEvaluationFrequency != 0) || (config_.statusReportFrequency != 0) ||
+      (config_.histogramsReportFrequency != 0) || (config_.metricsReportFrequency != 0)) {
+    periodicTimer_ = timers.add(std::chrono::seconds{config_.kDefaultBasePeriodicTimerIntervalSec},
+                                Timers::Timer::RECURRING,
+                                [this](Timers::Handle h) { this->doPeriodic(); });
+  } else {
+    LOG_WARN(logger_, "Periodic timer is disabled!" << KVLOG(KVLOG_PREFIX));
+  }
+
+  // Set/create subpool selector and purger
+  subPoolSelector_ =
+      subpoolSelector ? std::move(subpoolSelector) : std::make_unique<DefaultSubpoolSelector>(subpoolsConfig);
+  purger_ = purger ? std::move(purger) : std::make_unique<DefaultPurger>();
+
+  // initialize subpools buffer sizes vector
+  for (const auto& spc : subPoolsConfig_) subPoolsBufferSizes_.push_back(spc.bufferSize);
+  std::sort(subPoolsBufferSizes_.begin(), subPoolsBufferSizes_.end());
+
+  LOG_INFO(logger_,
+           "Done creating pool and allocating all Sub-Memory Pools:"
+               << KVLOG(KVLOG_PREFIX, subpoolsConfig.size()) << config_
+               << ". Subpool Selector:" << subPoolSelector_->name() << ". Purger:" << purger_->name() << stats_);
+}
+
+MultiSizeBufferPool::SubPool& MultiSizeBufferPool::getSubPool(uint32_t bufferSize) const {
+  auto iter = subpools_.find(bufferSize);
+  ConcordAssert(iter != subpools_.end());  // internal bug or memory corrupted - caller must validate
+  return *iter->second.get();
+}
+
+std::pair<char*, uint32_t> MultiSizeBufferPool::popBuffer(uint32_t requestedBufferSize) {
+  SubPool::AllocationResult result{};
+  uint32_t allocatedBufferSize{};
+  bool allocate{true};
+
+  DEBUG_PRINT(logger_, "pop enter:" << KVLOG(KVLOG_PREFIX));
+  /** 1st stage: repetitive, until reaching the largest subpool:
+   * 1a) Attempt to obtain a chunk from the target subpool that matches bufferSize.
+   * 1b) If such chinks is found - return chink and exit (success), If not continue to 1c.
```

Comment:
typo: chunk

Reply:
Thanks, addressed. Found another place with this typo..